primitives/src/pcs/univariate_kzg/mod.rs

```diff
@@ -347,6 +347,139 @@ impl<E: Pairing> UnivariatePCS for UnivariateKzgPCS<E> {
             .collect();
         Ok(evals)
     }
+
+    /// Input a polynomial, and multiple evaluation points,
+    /// compute a batch opening proof for the multiple points of the same
+    /// polynomial.
+    ///
+    /// Warning: don't use it when `points.len()` is large
+    fn multi_point_open(
+        prover_param: impl Borrow<<Self::SRS as StructuredReferenceString>::ProverParam>,
+        polynomial: &Self::Polynomial,
+        points: &[Self::Point],
+    ) -> Result<(Self::Proof, Vec<Self::Evaluation>), PCSError> {
+        if points.is_empty() {
+            return Err(PCSError::InvalidParameters(
+                "no point to evaluate and open".to_string(),
+            ));
+        }
+        let open_time = start_timer!(|| format!(
+            "Opening polynomial of degree {} at {} points",
+            polynomial.degree(),
+            points.len()
+        ));
+
+        let evals_time = start_timer!(|| "Computing polynomial evaluations");
+        let evals: Vec<Self::Evaluation> = points
+            .iter()
+            .map(|point| polynomial.evaluate(point))
+            .collect();
+        end_timer!(evals_time);
+
+        // Compute the polynomial \prod_i (X-point_i)
+        let vanish_poly =
+            Self::Polynomial::from_coefficients_vec(vec![-points[0], E::ScalarField::one()]);
+        let divisor: Self::Polynomial = points.iter().skip(1).fold(vanish_poly, |acc, point| {
+            &acc * &Self::Polynomial::from_coefficients_vec(vec![-*point, E::ScalarField::one()])
+        });
+
+        // Compute quotient poly
+        let witness_time = start_timer!(|| "Computing witness polynomial");
+        let witness_polynomial = polynomial / &divisor;
+        end_timer!(witness_time);
+
+        let (num_leading_zeros, witness_coeffs) =
+            skip_leading_zeros_and_convert_to_bigints(&witness_polynomial);
+
+        let proof: E::G1Affine = E::G1::msm_bigint(
+            &prover_param.borrow().powers_of_g[num_leading_zeros..],
+            &witness_coeffs,
+        )
+        .into_affine();
+
+        end_timer!(open_time);
+        Ok((Self::Proof { proof }, evals))
+    }
+
+    /// Verifies that `values` are the evaluation at the `points` of the
+    /// polynomial committed inside `comm`.
+    ///
+    /// Warning: don't use it when `points.len()` is large
+    fn multi_point_verify(
+        srs: &Self::SRS,
```

Comment:
We can also put `powers_of_h` into `VerifierParams` and replace `srs` with `vk`, but that involves a lot of code changes, and the `vk` will become larger.

Reply:
How about the following proposal:
1. Redefine `UnivariateVerifierParams` to have a field `powers_of_h[]`.
2. The API `extract_verifier_param(&self, _supported_degree: usize)` now will use `supported_degree` as a parameter to decide how many powers of h the verifier wants. (The parameter `_supported_deg` wasn't used before.) So now the user has the freedom to choose the vk size according to its need
3. Reuse `VerifierParams` in APIs

Reply:
Note for cleanness, we might remove `h` and `beta_h` from `VerifierParams` and use `powers_of_h[0/1]` instead. However, that could lead to many code changes. So I prefer to just add `extra_powers_of_h` and keep `h` and `beta_h`.

Reply:
I think I like this proposal. It's a good use of the existing API.

The only disadvantage is the user must still carry around powers of h even if he doesn't intend to check a multiproof. Astute users will know to set `supported_degree` low in this case but that's a pretty hacky use of the API. But I don't think this argument is sufficient to justify a more time-consuming solution. We can revisit later if there's demand.

I do like your suggestion for "cleanness". If it's not easy to remove then perhaps put a TODO in the comments or something.

Reply:
[70f0a0a](https://github.com/EspressoSystems/jellyfish/pull/406/commits/70f0a0a94e5c213707b889d1302a5eb0243dbaf8) and [7ad9865](https://github.com/EspressoSystems/jellyfish/pull/406/commits/7ad98656d4f843c61522392d0313624e957fe4a1) replaced `srs` with `vk` in the verify API.

Note that the vk also needs to includes some powers of g to enable commiting to evaluation poly I(X). 

For code cleanness, I've added an issue for removing fields `g`, `h`, `beta_h` in `VerifierParams` as those can be retrieved from `powers_of_g` and `powers_of_h`. It's hard to remove `g`, `h`, `beta_h` quickly as they have many appearances over the repo. (We should advocate exposing only APIs, not fields, which can make refactoring much easier in the future.): https://github.com/EspressoSystems/jellyfish/issues/439